cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,671 @@
+    auto const [value, did_overflow] = saturating<Op>{}(x.value(), y);
+    return {T{numeric::scaled_integer<Rep>{value, x.scale()}}, did_overflow};
+  }
+};
+
+/**
+ * @brief Functor to compute distance from current row for `unbounded` windows.
+ *
+ * An `unbounded` window runs to the boundary of the current row's group (inclusive).
+ *
+ * @tparam Grouping (inferred) type of object defining groups in the orderby column.
+ * @param groups object defining groups in the orderby column.
+ * @param direction direction of the window `PRECEDING` or `FOLLOWING`.
+ */
+template <typename Grouping>
+struct unbounded_distance_functor {
+  unbounded_distance_functor(Grouping groups, direction direction)
+    : groups{groups}, direction{direction}
+  {
+  }
+  Grouping const groups;
+  direction const direction;
+  [[nodiscard]] __device__ size_type operator()(size_type i) const noexcept
+  {
+    auto const row_info = groups.row_info(i);
+    if (direction == direction::PRECEDING) {
+      return i - row_info.group_start + 1;
+    } else {
+      return row_info.group_end - i - 1;
+    }
+  }
+};
+
+/**
+ * @brief Functor to compute distance from current row for `current_row` windows.
+ *
+ * A `current_row` window contains all rows whose value is the same as the current row.
+ *
+ * @tparam Grouping (inferred) type of object defining groups in the orderby column.
+ * @tparam OrderbyT (inferred) type of the entries in the orderby column
+ * @param groups object defining groups in the orderby column.
+ * @param direction direction of the window `PRECEDING` or `FOLLOWING`.
+ * @param order sort order of the orderby column.
+ * @param begin iterator to the begin of the orderby column.
+ */
+template <typename Grouping, typename OrderbyT>
+struct current_row_distance_functor {
+  current_row_distance_functor(Grouping groups,
+                               direction direction,
+                               order order,
+                               column_device_view::const_iterator<OrderbyT> begin)
+    : groups{groups}, direction{direction}, order{order}, begin{begin}
+  {
+  }
+  Grouping const groups;
+  direction const direction;
+  order const order;
+  column_device_view::const_iterator<OrderbyT> const begin;
+
+  [[nodiscard]] __device__ size_type operator()(size_type i) const noexcept
+  {
+    using Comp          = comparator_t<OrderbyT, current_row>;
+    auto const row_info = groups.row_info(i);
+    if (Grouping::has_nulls && i >= row_info.null_start && i < row_info.null_end) {
+      return direction == direction::PRECEDING ? i - row_info.null_start + 1
+                                               : row_info.null_end - i - 1;
+    }
+    if (direction == direction::PRECEDING) {
+      return 1 +
+             thrust::distance(
+               thrust::lower_bound(
+                 thrust::seq, begin + row_info.non_null_start, begin + i, begin[i], Comp{order}),
+               begin + i);
```

Comment:
Out of curiosity, could we check the compile-time/runtime tradeoff of moving this call into a non-inlined helper function and calling it both here and in the bounded functor? It could be bad for performance, but it's worth checking since thrust instantiations are often our biggest culprits.

Reply:
Outlining the thrust `lower_bound` calls results in 50% longer compile time (no change to the runtime)

Reply:
Wow OK sad.